# UPhO

UPhO find orthologs with and without in paralogs from input genefamily trees.
Addityinal scripts are proivided for a variety of task inclufing:
*minreID.py Rename sequence identifiers adding species (OTU) name
*blast_helper.sh generate all vs all blas search
*BlastResultCluste.py Cluster genes in gene families based on e vales threshold and a minimum number os OTUs.
*paMatrax.sh wrapper of gnu-parallelmaff, trimal and raxml (or fastree) for parallel esrimation of phylogenetic trees.
*UPhO.py The orthology evaluation tool. 
*GetFastafromRef.py Create fasta files from lists of sequence identifiers.
*Al2phylo.py A script to prepare MSA for phylogentic infernece where only one squence pero otu is retained and unique sequence identifiers removed.

Each script has its own  -help flag for detail on usage.

